docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+###  Detailed design : MinimalSerial
+
+The `MinimalSerial` class is a minimal abstract base class that provides basics I/O functionality. All other serial classes derive from it.
+
+**API description**
+
+This is the proposed API for the `MinimalSerial` class. Only two methods are provided to output `_base_putc()` and receive `_base_getc()` characters on the UART.
+
+```C
+class MinimalSerial : private NonCopyable<MinimalSerial> {
+protected:
+    MinimalSerial(PinName tx, PinName rx, int baud);
+    virtual ~MinimalSerial();
+    int _base_getc();
+    int _base_putc(int c);
+};
+```
+
+###  Detailed design : MinimalConsole
+
+`MinimalConsole` is an internal concrete class that derives from `MinimalSerial` and provides basic console functionality that can be used by the retarget code for resources constrained targets, see [Simplication of retarget code](#simplification-of-retarget-code).
+
+**API description**
+
+```C
+class MinimalConsole : public MinimalSerial, private NonCopyable<MinimalConsole> {
+public:
+    MinimalConsole(PinName tx, PinName rx, int baud);
+    virtual ~MinimalConsole();
+    int getc();
+    int putc(int c);
+};
+```
+
+The constructor provides arguments to create a serial port connected to the specified transmit and receive pins and set to the specified baud rate. If a target has serial support, by default a serial port is used for the console and the pins and settings for the port selection come from target header files and JSON settings. 
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+The below example shows pseudocode for how to print `Hello` to stdout using the new API. Printf cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.
+
+```C
+BufferedSerial serial(RX, TX, 115200);
+FILE *out = fdopen(&serial, "w");
+fprintf(out, "Hello");
+```
+and similarly for `UnbufferedSerial`
+
+```C
+UnbufferedSerial serial(RX, TX, 115200);
+FILE *out = fdopen(&serial, "w");
+fprintf(out, "Hello");
+```
+
+# Other Work
+
+## Related work-Simplification of retarget code
+
+In Mbed, retargeting is implemented by redefining the system I/O functions.
+The `FileHandle` abstract class provides an interface with file-like operations, such as read and write that can be redefined by the retarget code. The high-level library functions perform input/output by calling the low-level functions which use the system I/O functions to interface with hardware.
+
+The relationship of `FileHandle` and other APIs is as follows:
+
+![Retarget](./diagrams/retarget.png)
```

Comment:
In that diagram should probably also note `FileSystem` alongside `FileHandle`, which provides the `open` call.